### ISSUE TITLE
The new screen modes are unusable on MacOS

### ISSUE THREAD

**Issue by DBeitniz** [ppy/osu#28222]
### Type

Performance

### Bug description

The new version of Osu!lazer removed the option to use "Borderless" resolution, which was PERFECT. I have a lot of problem with the resolution now, its laggy, the performance of the game and input lag is worse, my Mac is using 20-30% more GPU and CPU and it just feels wrong. Also I have a program called "Bartender", which changes the menu bar, and it alters in with Osu and makes it actually terrible, I tried to use the 2560x1660 resolution and it still did that, with the Borderless option I had no issues at all, because the screen resolution the game used was 2560x1664 and the upper part where the menu bar is located was just black, which was perfect. I have some screenshots of this thing, please consider adding back the option for borderless resolution or whatever it was called. One thing to mention, before the update I could easily switch windows and go back to osu instantly, now if I go back to osu window it needs 5-8 seconds to scale for full screen, which is terrible and making me hate this game and installing back the version before. The first screenshot is of my screen without Osu (to show the menu bar which is on the top of the screen) and the other two screenshots are, how it looks in game, with the difference being one is "Default" and one is "2560x1600".

### Screenshots or videos

![SCR-20240519-lhjq](https://github.com/ppy/osu/assets/107348169/9ea40652-87bf-4416-9011-7305ce643f78)
![SCR-20240519-lhhm](https://github.com/ppy/osu/assets/107348169/8cdc928d-5a2a-471d-be25-6bbaa00230e2)
<img width="1470" alt="SCR-20240519-lhdc" src="https://github.com/ppy/osu/assets/107348169/d1f65132-a6f1-4d82-bd61-d1145d08c7fe">

### Version

2024.519.1

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/15369006/compressed-logs.zip)

**Comment by bdach** [comment]
Relevant: https://github.com/ppy/osu-framework/pull/6264

Will leave this to the macOS users to make a call on (cc @peppy @smoogipoo @frenzibyte)

**Comment by 45gfg9** [comment]
macOS +1. Actually the old "borderless" was conceptually closer to what is considered "fullscreen" (and vice versa), since it didn't just fill up the main display but instead, kept everything of osu!lazer in a clean, dedicated fullscreen window. And although the borderless mode is removed, the relevant keybindings and prompts are still present, and are causing confusion. Say, when I launched 2024.519.1, saw the windowed game and pressing Option+Enter repeatedly with no avail, then saw this in the configuration:
![c33c1759814268f3f5075dddbd6723c4](https://github.com/ppy/osu/assets/37043529/f58bbd09-46f3-491f-bd18-4bad21326cff)
And, to say the worst, makes the game somewhat unplayable. Is there a better workaround?

**Comment by S1n1st3rn3ss** [comment]
Adding to this, on Linux the situation is even worse

Changing resolutions and screen modes can cause a game crash and an monitor "loss" on x11, which I could reproduce with 100% guarantee on a tiling window manager with a multimonitor setup

Steps to repro:
1. Launch osu! in fullscreen
2. Change screen mode to Windowed
3. Change screen mode back to Fullscreen
4. Change resolution to any resolution below monitor's native
5. Change resolution back to native

At step 3, in-game resolution appears broken, 
![as can be seen here](https://github.com/ppy/osu/assets/49232762/35bb2f5f-c2f3-4279-828c-5ca835f31681), at step 5 the client crashes and the monitor it was on "disconnects", becoming an inactive xrandr output

Game logs:
[compressed-logs.zip](https://github.com/ppy/osu/files/15370627/compressed-logs.zip)

Xorg logs:
[log.log](https://github.com/ppy/osu/files/15370643/log.log)

System info (`inxi -F`):
[inxi.log](https://github.com/ppy/osu/files/15370664/inxi.log)

**Comment by frenzibyte** [comment]
Yes, open a separate issue thread.

**Comment by Vexcited** [comment]
Any update on this? Since the last update I can't no longer use the borderless mode neither...

**Comment by peppy** [comment]
@Vexcited fullscreen mode should work as a direct replacement for borderless mode. Could you please describe the specific issues you are having, along with video if relevant?

**Comment by Vexcited** [comment]
@peppy, it looks very broken when I alternate between my windows and osu! (when doing alt tabs for example)

https://github.com/user-attachments/assets/75e3b467-9721-49d2-8998-ff29c4ec2da3

I was very used to have osu! in a fullscreen space so I can simply switch to my desktop/apps quickly instead of having to minimize/alt tab osu! (which is kinda broken as you can see in the video)

I can't do this now without the bordered fullscreen option :

https://github.com/user-attachments/assets/2c598c20-5327-4bcd-8773-f15b4665b02f

Also, an issue you can't see here is that when I'm playing on my macbook, I have the notch at the top. It doesn't really matter in gameplay but in the menu, there's some buttons I can't see because the notch is on top of it.

**Comment by frenzibyte** [comment]
After a lengthy discussion over [discord](https://discord.com/channels/188630481301012481/589331078574112768/1309203145280454687), I want to discuss this issue at a different scale.

SDL3 provides three different window states on macOS:

- Windowed: You know, windowed.
- Fullscreen: The game removes its borders and its dimensions are set to match the display it’s displayed upon, achieving the effect of a fullscreen.
    - This is what osu! uses when you set the window mode to “Fullscreen” on SDL3 (not to be confused with SDL2, that uses something deprecated).
- Fullscreen space: The game acquires a dedicated space and becomes a fullscreen there, [completely using macOS fullscreen API](https://developer.apple.com/library/archive/documentation/General/Conceptual/MOSXAppProgrammingGuide/FullScreenApp/FullScreenApp.html).
    - This is equivalent to what you get when you run osu! on “Borderless” on SDL2, and now that I write this, I feel like some people already have their minds blown by this statement and what the mode internally does.
    - This mode is also used by three random macOS-supported games I have on my Steam library when I set them to “Fullscreen” mode. The games being [Moonlighter](https://store.steampowered.com/app/606150/Moonlighter/), [Celeste](https://store.steampowered.com/app/504230/Celeste/), and [Paper’s Please](https://store.steampowered.com/app/239030/Papers_Please/).
    - [Don’t Starve Together](https://store.steampowered.com/app/322330/Dont_Starve_Together/) does not use this mode, it currently uses the deprecated fullscreen mode that is what we had with “Fullscreen” mode on SDL2.

Right now on osu! we use two of three different window states on SDL3, that is “Windowed” and “Fullscreen”, and not “Fullscreen space” (although we are exposing it again as “Borderless” in https://github.com/ppy/osu-framework/pull/6264).

As per my previous comment, I have reported the major issues we have seen with the “Fullscreen” mode, but the last one touching on the window manager we use (yabai) has been mostly confirmed as wont-fix over SDL’s side. There are also other minor issues I have been seeing, and which I can still report upstream, but let me take a step back and say…

I want to shed light on the possibility of us dropping SDL3’s new “Fullscreen” mode, and fully replacing that mode with “Fullscreen space”, exposing it as fullscreen to the user, without borderless or anything else. This path is most friendly with macOS, as it uses what macOS provides for fullscreen mode (linking [this](https://developer.apple.com/library/archive/documentation/General/Conceptual/MOSXAppProgrammingGuide/FullScreenApp/FullScreenApp.html) again).

This implies that the user will not be able to select different display modes, as this mode does not offer such thing. I personally think of it as a necessary sacrifice, and any user who has been relying on such configuration will have to readjust. For any performance concerns, we will eventually add an option to change game resolution scale (see https://github.com/ppy/osu-framework/pull/5888) which can be used to accommodate for that.

Here's a side-to-side comparison showing behaviour of "Fullscreen" and "Fullscreen space":
 - "Fullscreen" / master: 
   
   https://github.com/user-attachments/assets/0c708e70-237c-43c1-89ab-a4493e1fe31b
 - "Fullscreen space" / my proposal:

   https://github.com/user-attachments/assets/133a12e0-eee3-472c-af62-77cf76bde5b8

I am testing this mode by checking out https://github.com/ppy/osu-framework/pull/6264 and applying the following diff:
```diff
diff --git a/osu.Framework/FrameworkEnvironment.cs b/osu.Framework/FrameworkEnvironment.cs
index 5b4b5ee8e..6ae0c9391 100644
--- a/osu.Framework/FrameworkEnvironment.cs
+++ b/osu.Framework/FrameworkEnvironment.cs
@@ -53,7 +53,7 @@ static FrameworkEnvironment()
             if (DebugUtils.IsDebugBuild)
                 AllowInsecureRequests = parseBool(Environment.GetEnvironmentVariable("OSU_INSECURE_REQUESTS")) ?? false;
 
-            UseSDL3 = RuntimeInfo.IsMobile || (parseBool(Environment.GetEnvironmentVariable("OSU_SDL3")) ?? false);
+            UseSDL3 = true;
         }
 
         private static bool? parseBool(string? value)
diff --git a/osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs b/osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs
index 8007a4a71..e49fa3302 100644
--- a/osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs
+++ b/osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs
@@ -150,7 +150,7 @@ public virtual IEnumerable<WindowMode> SupportedWindowModes
                 if (RuntimeInfo.IsMobile)
                     return new[] { Configuration.WindowMode.Fullscreen };
 
-                return Enum.GetValues<WindowMode>();
+                return new[] { Configuration.WindowMode.Windowed, Configuration.WindowMode.Borderless };
             }
         }
 
```

Note that this is not the final form of this proposal, the intention is that the "Fullscreen space" mode becomes used when choosing `WindowMode.Fullscreen`, and there will still be no support for `WindowMode.Borderless` which doesn't make sense on macOS anyway. Implementation details come later.

Oh yeah and I would like to give a huge thanks to Safari for having me write this twice thanks to some memory overflow, productivity is totally off the roof.

**Comment by peppy** [comment]
TO confirm, we're waiting on a subsequent PR with the full changes you propose (using fullscreen space)?

**Comment by frenzibyte** [comment]
I'm waiting on whether you agree with what I'm saying above before I go ahead with it.

I suppose you understand that what I'm essentially proposing here is just a couple of SDL calls done on o!f's SDL3Window implementation and that's it.

**Comment by peppy** [comment]
there's a lot of text, but you seem to be proposing "fullscreen space" or "window" and no other options?

**Comment by frenzibyte** [comment]
Correct. These blocks of text were meant to be to get you in the loop of what I'm seeing for macOS going forward, I hope my points came across to you 😅

**Comment by frenzibyte** [comment]
Deprioritising and unassigning as the "fullscreen space" mode proposed above will be brought back on SDL3 starting next release under the `Borderless` window mode, and the issues highlighted by this thread in https://github.com/ppy/osu/issues/28222#issuecomment-2488914891 have been resolved for the most part, only requiring an SDL3 bump to be included in the game.

The end goal of my proposal is to drop the other fullscreen mode as well, but I'll just leave that be as it involves dropping display modes support entirely and that's probably not worth it (until at least we have support for downscaling game resolution).
